src/devices/cmr113.c

```diff
@@ -0,0 +1,110 @@
+/** @file
+    CMR112 Clipsal cent-a-meter power meter
+
+    Copyright (C) 2021 Michael Neuling <mikey@neuling.org>
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+*/
+/*
+*/
+
+#include "decoder.h"
+
+#define COMPARE_BITS 83
+#define COMPARE_BYTES (COMPARE_BITS/8)
+
+static int cmr113_decoder(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    int start, bit;
+    uint8_t buf[4];
+    uint8_t b1[COMPARE_BYTES], b2[COMPARE_BYTES];
+    bitbuffer_t b = {0};
+    double current, power;
+    data_t *data;
+
+    if ((bitbuffer->bits_per_row[0] < 350) || (bitbuffer->bits_per_row[0] > 450))
+	    return DECODE_ABORT_LENGTH;
+
+    bitbuffer_extract_bytes(bitbuffer, 0, 0, buf, 32);
+    if ((buf[0] != 0xb0) || (buf[1] != 0x00) || (buf[2] != 0x00))
+	    return DECODE_ABORT_EARLY;
+    /*
+     * Modulation is pretty strange
+     * It's PWM where you look at all pulses high and low.
+     *   LONG LONG means no toggle
+     *   LONG SHORT SHORT means toggle
+     * Code below takes the raw pulse data in stream bitbuffre and
+     * turns it into a new data b;
+     */
+    start = 0;
+    bit = 0;
+    bitbuffer_clear(&b);
+    while ((start + 3) < bitbuffer->bits_per_row[0]) {
+	    bitbuffer_extract_bytes(bitbuffer, 0, start, buf, 3);
+	    if ((buf[0] >> 6) == 0x00) { // top two bits are 0b00 = no toggle
+		    start += 2;
+		    bitbuffer_add_bit(&b, bit);
+	    } else if ((buf[0] >> 5) == 0x03) { // top two bits are 0b011 = toggle
+		    start += 3;
+		    bit = 1 - bit; // toggle
+		    bitbuffer_add_bit(&b, bit);
+	    } else if (start == 0)
+		    start += 1; /* first bit doesn't decode */
+	    else
+		    /* we don't have enough bits */
+		    return DECODE_ABORT_LENGTH;
+    }
+    if (b.bits_per_row[0] < 2*COMPARE_BITS + 2)
+	    return DECODE_ABORT_LENGTH;
+    /* Compare the repeated section to ensure data integrity */
+    bitbuffer_extract_bytes(&b, 0, 0, b1, COMPARE_BITS);
+    bitbuffer_extract_bytes(&b, 0, COMPARE_BITS+2, b2, COMPARE_BITS);
+    if (memcmp(b1, b2, COMPARE_BYTES) != 0)
+	    return DECODE_FAIL_MIC;
+
+    /* Data is all good, so extract 3 phases of current */
+    current = 0;
+    for (int i = 0; i < 3 ; i++){
+	    bitbuffer_extract_bytes(&b, 0, 36 + i*10, buf, 10);
+	    reflect_bytes(buf, 2);
+	    current += ((float)buf[0] + ((buf[1] & 0x3) << 8))/10.0;
+    }
+    /* These seem to be exclusively in Australia, so assume 240V */
+    power = 240 * current;
+
+    data = data_make(
+	    "model",   "",         DATA_STRING, "CMR112",
+	    "power_W", "Power",    DATA_FORMAT, "%.0f W", DATA_DOUBLE,  power,
+	    "current", "Current",  DATA_FORMAT, "%.1f A", DATA_DOUBLE,  current,
```

Comment:
SI unit missing.

Reply:
SI missing on the current? should be "current_A" ?

Reply:
I think that is the correct unit.